### PR TITLE
Include PCRE, Grok + libdict in the repo

### PR DESCRIPTION
**Pull request by actgardner**
This PR compares all of the PCRE, Grok, libdict and original Go code to the final modified version. It's the diff between dumping all the source files into one big directory, and a version that actually builds on all the platforms we need.

The end result is a version that builds on Mac and Linux with no external dependencies - PCRE and Grok are vendored in. I wish there was a tidier way to arrange the source, it's all dumped in one directory.

On Windows we need mingw-w64 and libportablexdr, there's no getting around that.

There are a few large deletions from grok, I took them out because it was too much code to wade through.

### CODE REVIEW

File: cgrok/tree.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#include <stdbool.h>
```
Comment (actgardner) [inline cgrok/tree.h:1]: Header file for the shim that connects grok to libdict. This implements all the tokyocabinet functionality we need.

File: cgrok/tree.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
+#include "tree.h"
```
Comment (actgardner) [inline cgrok/tree.c:1]: The shim that allows libdict to replace tokyocabinet. I wrote this from scratch, focus here :)

edit: The tokyocabinet spec might be helpful: http://fallabs.com/tokyocabinet/spex-en.html

File: cgrok/pcreposix.c [hunk 3]
```diff
@@ -57,6 +57,7 @@ previously been set. */
 #  define PCREPOSIX_EXP_DEFN __declspec(dllexport)
 #endif
 
+#define PCRE_STATIC
```
Comment (actgardner) [inline cgrok/pcreposix.c:61]: Necessary on Windows to avoid the exported names getting mangled for linking against a DLL.

File: cgrok/pcregrep.c [hunk 4]
```diff
@@ -1,2379 +0,0 @@
-/*************************************************
```
Comment (actgardner) [inline cgrok/pcregrep.c]: We don't need a whole grep program.

File: cgrok/pcre_internal.h [hunk 5]
```diff
@@ -310,7 +310,7 @@ defined here.
 The macros are controlled by the value of LINK_SIZE. This defaults to 2 in
 the config.h file, but can be overridden by using -D on the command line. This
 is automated on Unix systems via the "configure" command. */
-
+#define LINK_SIZE 2
```
Comment (actgardner) [inline cgrok/pcre_internal.h]: It was necessary to set this here, it didn't seem to work when it was exported in `config.h`. Not 100% sure why.

File: cgrok/pcre_chartables.c [hunk 6]
```diff
@@ -0,0 +1,198 @@
+/*************************************************
```
Comment (alanctgardner) [inline cgrok/pcre_chartables.c]: Include the default tables that are normally built by the Makefile.

File: cgrok/grok_pattern.c [hunk 7]
```diff
@@ -76,10 +76,14 @@ int grok_patterns_import_from_file(const grok_t *grok, const char *filename) {
   memset(buffer, 0, filesize);
   bytes = fread(buffer, 1, filesize, patfile);
   if (bytes != filesize) {
-    grok_log(grok, LOG_PATTERNS, "Unable to open '%s' for reading: %s",
-             filename, strerror(errno));
-    fprintf(stderr, "Expected %zd bytes, but read %zd.", filesize, bytes);
-    return GROK_ERROR_UNEXPECTED_READ_SIZE;
+    // Using ftell on a text file, we may get less bytes than we expect.
```
Comment (actgardner) [inline cgrok/grok_pattern.c:79]: On Windows in particular, opening a file with "r" and using `ftell` to get the length isn't 100% accurate - there may be fewer characters, but never more. Since it gives an upper bound I rewrote this to check for EOF, and error only if we didn't read the whole file.

File: cgrok/grok.c [hunk 8]
```diff
@@ -1,7 +1,4 @@
 #include "grok.h"
-#include <dlfcn.h>
```
Comment (actgardner) [inline cgrok/grok.c]: Including `dlfcn.h` on Windows is a nightmare, and we will never be `dlopened` so I ripped this out.

File: cgrok/grok_discover.c [hunk 9]
```diff
@@ -186,6 +186,8 @@ void grok_discover(const grok_discover_t *gdt, /*grok_t *dest_grok, */
     if (max_matchlen == 0) { /* No valid matches were found */
       if (first_match_endpos > 0) {
         offset += first_match_endpos;
+      } else {
```
Comment (actgardner) [inline cgrok/grok_discover.c:189]: This resolves a super confidence-inspiring issue that was fixed in master but has yet to make it into any release: https://github.com/jordansissel/grok/commit/25acb06deae35072d530285f505a1a428aedf734

**Comment by brianc** [comment]
Kudos on the mighty C work.  It's always kinda cool to read C and go "oh so _thats_ how computers work."  Seriously though - Might be worth it to write some tests for your tree implementation?

**Comment by brianc** [comment]
I wonder if tokyo cabinet itself has some tests? Maybe they'd work to just copy/paste some over?

File: cgrok/tree.c [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  void *val = tcpack(vbuf, vsiz);
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
```
Comment (brianc) [inline cgrok/tree.c:93]: Should you check anything on the `inserted` value after doing an insert for possible failures? I'm assuming that's what its used for.

**Comment by actgardner** [comment]
@brianc - it would be awesome to use the tokyocabinet unit tests, but TC is LGPL licensed. I haven't looked at any of the source just in case.

File: cgrok/tree.c [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+  }
+  
+  TCLISTELEM *removed = head->next;
+  head->next = removed->next;
+  void* elem = removed->elem;
+  *sp = removed->size;
+  free(removed);
+  list->len -= 1;
+  return elem;
+}
```
Comment (brianc) [inline cgrok/tree.c:235]: This removes an element from the list & returns a pointer to the element? So its up to the caller to `free` the element even though the list internally does a `malloc` to store the element?

Comment (actgardner) [inline cgrok/tree.c:235]: The TC docs are a big ball of crazy. This is what it says about `tclistremove`: 

```
Because an additional zero code is appended at the end of the region of the return value, the return value can be treated as a character string. 
Because the region of the return value is allocated with the `malloc' call, it should be released with the `free' call when it is no longer in use. 
If `index' is equal to or more than the number of elements, no element is removed and the return value is `NULL'.
```

I take this to mean "When you give us a value, we malloc sizeof(value)+1 and copy your value there with a NULL terminator. When you ask for it back we give you our copy, and freeing it is your problem. " But I could be wrong - I used the same pattern for the values in the tree as well.

File: cgrok/tree.c [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+  if (index >= list->len) {
+    // Out of bounds, do nothing
+    return;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i <= index; i++) {
+    head = head->next;
+  }
+  free(head->elem);
+  head->elem = ptr;
```
Comment (brianc) [inline cgrok/tree.c]: On line 136 you call `tclistpack` on the supplied pointer to the new element before assigning it to the `elem` pointer but you don't call it here when overwriting an existing element so if you update an element & pull it back out it wont be null terminated.

File: cgrok/tree.c [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+  for (int i=0; i <= index; i++) {
+    head = head->next;
+  }
+  *sp = head->size;
+  return head->elem;
+}
+
+void tclistdel(TCLIST *list) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < list->len; i++) {
```
Comment (brianc) [inline cgrok/tree.c]: nit: On line 132 you surround the `list->len` with parenthesis. Not sure if that was a style thing that didn't come through here or there's some subtle bug you were trying to avoid.

File: cgrok/tree.c [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+void tclistover(TCLIST *list, int index, const void *ptr, int size) {
+  if (index >= list->len) {
+    // Out of bounds, do nothing
+    return;
+  }
+  TCLISTELEM *head = list->head;
+  for (int i=0; i <= index; i++) {
+    head = head->next;
+  }
+  free(head->elem);
```
Comment (brianc) [inline cgrok/tree.c]: On line 190 you check `head->elem` for `NULL` before calling free on it. No such behavior here.

File: cgrok/tree.c [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+  list->head = newElem;
+  return list;
+}
+
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
```
Comment (brianc) [inline cgrok/tree.c]: My C is terrible, so I could be wrong & you don't have to educate me but I thought this could return an error code indicating there was no way to allocate anymore memory & you were supposed to check that error code & then do some logging & crash your process?  I know its rare for malloc to fail, but..hay.

File: cgrok/tree.c [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop) {
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  tree->dict = hb_dict_new(cp, tcfree);
+  tree->iter = NULL;
+  return tree; 
+}
+
+// Iterate over elements in ascending key order
+void tctreeiterinit(TCTREE *tree) {
+  if (tree->iter != NULL) {
```
Comment (brianc) [inline cgrok/tree.c:46]: I guess tokyo cabinet isn't threadsafe?  This would be pretty easy for one to iterate 1/2 way through the dict, then someone else starts iterating and then the first person returns to further iterate only to find their iteration point has been reset?

File: cgrok/tree.c (hunk 10 above)
Comment (actgardner) [inline cgrok/tree.c:93]: `dict_insert` either makes a new node in the tree if the key doesn't exist, or returns the existing node. Either way, `tctreeput` always overwrites the existing value. `tctreeputkeep` checks `inserted` and doesn't overwrite existing values.

edit: Duh, need to free the old value.

**Comment by brianc** [comment]
I put a few questions down on a few things - I shouldn't be the final authority on this PR though. I've not done enough prod C code to know any of the landmines.  One thing I think will help though is putting some comments into your tokyo shim. You have a ton of good info in your head around implementing it and the decisions you made, expected behaviors from tokyo cabinet, etc that could be lost if you return to this code 6 months to a year from now at 1 in the morning because something is segfaulting at some customer running Windows 2003 server.

File: cgrok/tree.c (hunk 15 above)
Comment (actgardner) [inline cgrok/tree.c]: Yeah, malloc can return NULL if we run out of memory or address space. I've never heard a really satisfying answer to how to handle this, anyone have any ideas? The process would crash instantly at the `memcpy` to NULL right now.

File: cgrok/tree.c (hunk 13 above)
Comment (actgardner) [inline cgrok/tree.c]: Just style, removed the parens in both spots for consistency.

File: cgrok/tree.c (hunk 12 above)
Comment (actgardner) [inline cgrok/tree.c]: Good catch!

File: cgrok/tree.c (hunk 15 above)
Comment (brianc) [inline cgrok/tree.c]: Yeah beyond my experience level. I've seen people use `ASSERT` there which indeed crashes the program, but at least it tells you why.

File: cgrok/tree.c (hunk 16 above)
Comment (actgardner) [inline cgrok/tree.c:46]: Yeah, that's their bad design choice. `tciterinit` and `tctreeiternext` both take a `TCTREE` as an argument, there's no way to have multiple iterators on the same tree at the same time.

File: cgrok/tree.c (hunk 15 above)
Comment (adamlamar) [inline cgrok/tree.c]: It depends on who you ask. If `malloc` fails, the process (and possibly the entire system) is probably under a lot of pressure anyway, and only a carefully designed logging subsystem would actually work under those conditions.

Perhaps an improvement would be to check if any of the `malloc` calls return `NULL` and return `NULL` from `tclistnew` on that condition? Then you're just punting the problem to someone else, but I believe that is more correct design.

Comment (actgardner) [inline cgrok/tree.c]: I can do that in some places (propagate the NULL up), but in methods like `tctreeput` we have no way to signal an error to the user. It seems like failing fast (with an assert like @brianc mentioned) would be better than silently continuing without having inserted the value.

Comment (adamlamar) [inline cgrok/tree.c]: @alanctgardner Absolutely agree, though I'm unsure about using `assert` in production code because it doesn't give the user of the code any chance to recover. If you have no way to signal an error, then I guess the API is busted and you don't have a choice really. 

I think the classic strategies are to wait and see if time will resolve the `malloc` problem (e.g. try again soon, fail after some number of attempts), or to drop some memory caches if you have them.

File: test.c [hunk 17]
```diff
@@ -0,0 +1,72 @@
+#include "tree.h"
```
Comment (actgardner) [inline test.c:1]: It's not possible to valgrind Go binaries, so I wrote this program and caught a few memory leaks in `tree.c` with it. We are now leakless.

**Comment by joey** [comment]
> mingw-w64

Do we need to add this to our README? Do we have Windows build instructions in our README?

File: cgrok/grok.go [hunk 18]
```diff
@@ -1,8 +1,9 @@
 package grok
 
 /*
-#cgo LDFLAGS: -lgrok
-#include <grok.h>
+#cgo CFLAGS: -I. -std=gnu99
+#cgo windows LDFLAGS: -L. -lportablexdr -lws2_32
+#include "grok.h"
```
Comment (joey) [inline cgrok/grok.go:6]: I'm not familiar with what go is doing here, but it looks like it's providing some flags for the cgo tool. Does this change the location where it looks for `grok.h` from the system include directories to a relative directory?

File: cgrok/grok.h [hunk 19]
```diff
@@ -4,8 +4,10 @@
 #ifndef _GROK_H_
 #define _GROK_H_
 
+#include <stdio.h>
+#include "tree.h"
```
Comment (joey) [inline cgrok/grok.h:8]: I assume this is your non-tokyo cabinet tree implementation?

File: cgrok/grok_discover.c (hunk 9 above)
Comment (joey) [inline cgrok/grok_discover.c:189]: Do we also need the change to `discover_main.c` that's included in that upstream commit?

File: cgrok/grok_pattern.c (hunk 7 above)
Comment (joey) [inline cgrok/grok_pattern.c:79]: Is there any potential risk with the buffer being a larger size than the number of bytes that we read?

File: cgrok/grok_pattern.c [hunk 20]
```diff
@@ -76,10 +76,14 @@ int grok_patterns_import_from_file(const grok_t *grok, const char *filename) {
...
+    // Test if we hit EOF
+    if (!feof(patfile)) {
+      grok_log(grok, LOG_PATTERNS, "Unable to open '%s' for reading: %s",
```
Comment (joey) [inline cgrok/grok_pattern.c]: nit: This log message isn't accurate as the file was opened above.

File: cgrok/grok.c [hunk 21]
```diff
@@ -79,53 +73,3 @@ void grok_clone(grok_t *dst, const grok_t *src) {
   dst->logmask = src->logmask;
   dst->logdepth = src->logdepth + 1;
 }
-
-static int grok_pcre_callout(pcre_callout_block *pcb) {
```
Comment (joey) [inline cgrok/grok.c]: Is this related to removing the `dlopen` functionality?

File: cgrok/grok.h [hunk 22]
```diff
@@ -4,8 +4,10 @@
...
+#define PCRE_STATIC
```
Comment (joey) [inline cgrok/grok.h:10]: nit: probably worth the `PCRE_STATIC` comment you include in other places

File: cgrok/grokre.c [hunk 23]
```diff
@@ -217,7 +219,7 @@ static char *grok_pattern_expand(grok_t *grok) {
       if (definition_len > 0) {
         /* We got an in-line definition */
         /* discard const-ness with reckless abandon! */
-        pattern_regex = strndup(full_pattern + capture_vector[g_cap_definition * 2], definition_len);
+        pattern_regex = string_ndup(full_pattern + capture_vector[g_cap_definition * 2], definition_len);
```
Comment (joey) [inline cgrok/grokre.c:222]: I couldn't find docs on `string_ndup`, why did this change?

File: cgrok/grok.go (hunk 18 above)
Comment (actgardner) [inline cgrok/grok.go:6]: `-I.` should look in the local directory for headers. According to https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html GCC will look at the -I _before_ the system paths, so we've got that covered.

File: cgrok/tree.c [hunk 24]
```diff
@@ -0,0 +1,275 @@
...
+
+// A shim to use dictlib trees instead of TC
+struct TCTREE {
+  dict *dict;
+  dict_itor *iter;
+};
+
+// Toyko cabinet accepts any arbitrary-sized data as a tree key.
+// We support this by using the first 4 bytes of the key to store the length.
+// So we have to wrap the libdict comparators with implementations that skip the first four bytes
```
Comment (joey) [inline cgrok/tree.c:8]: Does `libdict` use only 4 byte integers? If not, how does `dict_int_cmp` know the size of the keys?

File: cgrok/grok.h (hunk 19 above)
Comment (actgardner) [inline cgrok/grok.h:8]: Yes. I'm at a loss for a better name, "Osaka Cupboard"?

File: cgrok/grok.c (hunk 21 above)
Comment (actgardner) [inline cgrok/grok.c]: Yes. `grok_pcre_callout` was only used when Ruby dlopened this library.

File: cgrok/tree.c [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+// of the key
+int tccmpint32(const void* k1, const void* k2) {
+  return dict_int_cmp(k1+4, k2+4);
+}
+
+int dict_var_str_cmp(const void* k1, const void* k2) {
+  return dict_str_cmp(k1+4, k2+4);
```
Comment (joey) [inline cgrok/tree.c:17]: Are these null terminated strings?

File: cgrok/grokre.c (hunk 23 above)
Comment (actgardner) [inline cgrok/grokre.c:222]: I brought in `string_ndup` from the `grok` Github master, Windows doesn't have `strndup` so they reimplemented it as `string_ndup`. It's part of (pattern) compilation and not on the critical path so I'm not as worried about speed.

File: cgrok/tree.c [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+void tcfree(void *key, void *value) {
+  free(key);
+  free(value);
+}
+
+TCTREE *tctreenew(void) {
+  return tctreenew2(dict_var_str_cmp, 0);
```
Comment (joey) [inline cgrok/tree.c]: nit: You should pass an explicit `NULL` instead of `0`

File: cgrok/tree.c [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+// Make a new tree with a user-defined comparator (must be tccmpint32 or dict_var_str_cmp),
+// because of how we pack the keys
+TCTREE *tctreenew2(dict_compare_func cp, void *cmpop) {
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  if (!tree) {
+    return NULL;
```
Comment (joey) [inline cgrok/tree.c]: Does the grok code do the right thing if we return `NULL` here?

File: cgrok/tree.c (hunk 24 above)
Comment (actgardner) [inline cgrok/tree.c:8]: I just checked, `libdict` just uses `unsigned int` and not a standard int type :( I'll reimplement it for `uint32_t` specifically.

File: cgrok/tree.c (hunk 25 above)
Comment (actgardner) [inline cgrok/tree.c:17]: Always. As part of the TC API we stick an extra null on everything, just for fun.

File: cgrok/tree.c (hunk 27 above)
Comment (actgardner) [inline cgrok/tree.c]: No, grok doesn't actually check the return value at all. This was part of the larger discussion of how we should handle malloc failing.

File: cgrok/tree.c [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+  }
+  return NULL;
+}
+
+// Pack a key or value: uint32_t + body + NULL
+// Callers might use integers, structs, etc. for 
+// keys and values, so we can't depend on null-termination
+// to find the lenght of a key.
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
```
Comment (joey) [inline cgrok/tree.c]: `buf` should be a `void *` to match the return value of the function.

File: cgrok/tree.c [hunk 29]
```diff
@@ -0,0 +1,275 @@
...
+  if (!buf) {
+    return NULL;
+  }
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
```
Comment (joey) [inline cgrok/tree.c]: The last byte should be set to `'\0'` rather than `0`.

File: cgrok/tree.c [hunk 30]
```diff
@@ -0,0 +1,275 @@
...
+  return buf;
+}
+
+// Insert a key-value pair. If the key already exists the value will be overwritten
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+
```
Comment (joey) [inline cgrok/tree.c]: Missing error handling code?

File: cgrok/tree.c [hunk 31]
```diff
@@ -0,0 +1,275 @@
...
+// Insert a key-value pair. If the key already exists the value will be overwritten
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    
+  }
+  void *val = tcpack(vbuf, vsiz);
+  if (!val) {
+
```
Comment (joey) [inline cgrok/tree.c]: Missing error handling code?

File: cgrok/tree.c [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  } 
+  void *val = tcpack(vbuf, vsiz); 
+  if (!val) {
+    fprintf(stderr, "Failed to malloc tree value\n");
+    return;
```
Comment (joey) [inline cgrok/tree.c]: If this code path happens, the `key` is never freed.

File: cgrok/tree.c [hunk 33]
```diff
@@ -0,0 +1,275 @@
...
+
+// Remove all elements from the tree
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
```
Comment (joey) [inline cgrok/tree.c]: `if (tree->iter != NULL)`

File: cgrok/tree.c [hunk 34]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+// Remove all elements from the tree
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
```
Comment (joey) [inline cgrok/tree.c:151]: Do we need to do defensive `NULL` checks for our parameters?

File: cgrok/tree.c [hunk 35]
```diff
@@ -0,0 +1,275 @@
...
+  TCLIST *list = malloc(sizeof(TCLIST));
+  if (!list) {
+    return NULL;
+  }
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  if (!newElem) {
+    return NULL;
+  }
+  newElem->elem = 0;
```
Comment (joey) [inline cgrok/tree.c]: Assuming `elem` is a pointer, it should be set to `NULL` rather than `0`.

File: cgrok/tree.c [hunk 36]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+// Null-terminate the value to be inserted, as TokyoCabinet does
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
+  if (!elem) {
+    return NULL;
+  }
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
```
Comment (joey) [inline cgrok/tree.c]: You should use `'\0'` instead of `0`.

File: cgrok/tree.c [hunk 37]
```diff
@@ -0,0 +1,275 @@
...
+// Append a new element at the end of the list
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < list->len; i++) {
+    head = head->next;
+  }
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = tclistpack(ptr, size);
+  if (!newElem->elem) {
+    fprintf(stderr, "Failed to malloc list element\n");
```
Comment (joey) [inline cgrok/tree.c]: If this code path is followed, we'll leak the `newElem` that we allocated above.

File: cgrok/tree.c [hunk 38]
```diff
@@ -0,0 +1,275 @@
...
+  return elem;
+}
+
+// Append a new element at the end of the list
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < list->len; i++) {
+    head = head->next;
+  }
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
```
Comment (joey) [inline cgrok/tree.c]: Check `malloc` for `NULL`

File: cgrok/tree.c [hunk 39]
```diff
@@ -0,0 +1,275 @@
...
+  newElem->elem = tclistpack(ptr, size);
+  if (!newElem->elem) {
+    fprintf(stderr, "Failed to malloc list element\n");
+    return;
```
Comment (joey) [inline cgrok/tree.c]: Do we need to fail fast here? There doesn't' seem to be an error code returned from this function?

File: cgrok/tree.c [hunk 40]
```diff
@@ -0,0 +1,275 @@
...
+    return NULL;
+  }
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
+  return elem;
+}
+
+// Append a new element at the end of the list
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
```
Comment (joey) [inline cgrok/tree.c]: nit: This variable is only the `head` of the list for a short while. Something like `item` or `cur` would be better.

File: cgrok/tree.c [hunk 41]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+// Remove the element at index, and return a pointer to it.
+// The caller is responsible for freeing the returned element.
+// Returns NULL if index is out of bounds.
+void *tclistremove(TCLIST *list, int index, int *sp) {
+  if (index < 0 || index >= list->len) {
+    return NULL;
+  }
+  TCLISTELEM *head = list->head;
```
Comment (joey) [inline cgrok/tree.c]: nit: Using `head` for a current item pointer

File: cgrok/tree.c [hunk 42]
```diff
@@ -0,0 +1,275 @@
...
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
```
Comment (joey) [inline cgrok/tree.c:158]: From my quick read of the TC spec `TCLIST` is an array list rather than a linked list. Does this change any performance assumptions of grok's use of `TCLIST`?

File: cgrok/tree.h [hunk 43]
```diff
@@ -0,0 +1,44 @@
...
+typedef struct TCLISTELEM TCLISTELEM;
+
+typedef struct {
+  TCLISTELEM *head;
+  int len;
+} TCLIST;
+
+struct TCLISTELEM {
+  int size;
+  void *elem;
```
Comment (joey) [inline cgrok/tree.h]: nit: I'd call this `val` as `elem` confuses the data in the list with an element of the list. Or maybe the struct should be called `TCLISTNODE` as that's the terminology that I'm more used to.

File: cgrok/grok.h (hunk 19 above)
Comment (MarkRoddy) [inline cgrok/grok.h:8]: Nice one

edit: formatting

File: cgrok/tree.c [hunk 44]
```diff
@@ -0,0 +1,275 @@
+#include "tree.h"
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+// A shim to use dictlib trees instead of TC
+struct TCTREE {
```
Comment (joey) [inline cgrok/tree.c]: Why is this in the `.c` instead of the `.h`?

File: cgrok/tree.c (hunk 25 above)
Comment (joey) [inline cgrok/tree.c:17]: My question was more if `dict_str_cmp` expects null terminated strings.

**Comment by joey** [comment]
IMO, if grok doesn't handle the TC code returning `NULL`, then we either have to take that on as technical debt for ourselves to fix in our fork or we should fail fast.

**Comment by joey** [comment]
A couple of open questions/nits. Let me know when this is ready for another pass.

File: cgrok/grok_discover.c (hunk 9 above)
Comment (actgardner) [inline cgrok/grok_discover.c:189]: We don't need `discover_main.c` for the library functionality, so I already pulled it out of this repo.

File: cgrok/tree.c (hunk 25 above)
Comment (actgardner) [inline cgrok/tree.c:17]: Yes, `dict_str_cmp` expects null-terminated strings.

File: cgrok/grok_pattern.c (hunk 7 above)
Comment (actgardner) [inline cgrok/grok_pattern.c:79]: On line 76 we `memset` the whole buffer with NULLs, this is safe.

File: cgrok/tree.c (hunk 44 above)
Comment (actgardner) [inline cgrok/tree.c]: Moved to the `.h` file, just me being sloppy.

File: cgrok/tree.c (hunk 34 above)
Comment (actgardner) [inline cgrok/tree.c:151]: I added one, `dict_free` fails an assertion if the argument is NULL.

File: cgrok/tree.c (hunk 42 above)
Comment (actgardner) [inline cgrok/tree.c:158]: The only place we use `TCLIST` is to store multiple matches with the same name in a single node of the tree:

```
  listsize = tclistnum(by_name_list);
  for (i = 0; i < listsize; i++) {
    grok_capture *list_gct;
    list_gct = (grok_capture *)tclistval(by_name_list, i, &unused_size);
    if (list_gct->id == gct->id) {
      tclistremove(by_name_list, i, &unused_size);
      break;
    }
  }
  tclistpush(by_name_list, gct, sizeof(grok_capture));
  tctreeput(grok->captures_by_name, gct->name, gct->name_len, by_name_list, sizeof(TCLIST));
```

It has some unfortunate runtime complexity (iterating over every element is going to be `O(nlogn)` with a linked list), and `tclistpush` is `O(n)`. Typically `n` will be less than 8, so I'm not sure if it's worth optimizing for.

**Comment by actgardner** [comment]
@joey This is ready for a second pass. I went ahead with fail-fast error handling for `malloc`:

On `malloc` failure we try to print an error to `stderr`, and call `exit()`. I did some testing, an `exit()` in cgo causes the whole program to stop silently (including any goroutines). `abort()`, or a failed assertion, has a similar effect but the Go VM will trap the signal and print a stack trace as well. I don't think we want a stack trace or core dump in this scenario.

edit: As far as Windows build instructions go, there's a Google doc that describes how to set up our Windows build environment. I've updated it with the new dependencies.

**Comment by joey** [comment]
nit: it would be nice to make sure our error messages when malloc fails are verbose enough for us to find where it failed. 

LGTM, +1

**Comment by actgardner** [comment]
Merged here: 2b374422853a1032716aee98b60ce7d02423f59c
